Fix GH-29442 DataFrame.groupby doesn't preserve _metadata

Japanuspus opened the pull request:
This bug is a regression in v1.1.0 and was introduced by the fix for GH-34214 in commit [6f065b].

Underlying cause is that the `*Splitter` classes do not use the `._constructor` property and do not call `__finalize__`.

Please note that the method name used for `__finalize__` calls was my best guess since documentation for the value has been hard to find.

[6f065b]: https://github.com/pandas-dev/pandas/commit/6f065b6d423ea211d803e8be93c27f547541c372

- [x] closes #29442
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Japanuspus committed:
Fix GH-29442 DataFrame.groupby doesn't preserve _metadata

This bug is a regression in v1.1.0 and was introduced by the fix for GH-34214 in commit [6f065b].

Underlying cause is that the `*Splitter` classes do not use the `._constructor` property and do not call `__finalize__`.

Please note that the method name used for `__finalize__` calls was my best guess since documentation for the value has been hard to find.

[6f065b]: https://github.com/pandas-dev/pandas/commit/6f065b6d423ea211d803e8be93c27f547541c372

pep8speaks commented:
Hello @Japanuspus! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/tests/groupby/test_custom_metadata.py`](https://github.com/pandas-dev/pandas/blob/7cc4d5375e57277f316c5a8bb17a58ab8003c5f1/pandas/tests/groupby/test_custom_metadata.py):

> [Line 4:89](https://github.com/pandas-dev/pandas/blob/7cc4d5375e57277f316c5a8bb17a58ab8003c5f1/pandas/tests/groupby/test_custom_metadata.py#L4): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 88 characters)
> [Line 5:89](https://github.com/pandas-dev/pandas/blob/7cc4d5375e57277f316c5a8bb17a58ab8003c5f1/pandas/tests/groupby/test_custom_metadata.py#L5): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 88 characters)
> [Line 5:99](https://github.com/pandas-dev/pandas/blob/7cc4d5375e57277f316c5a8bb17a58ab8003c5f1/pandas/tests/groupby/test_custom_metadata.py#L5): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 7:89](https://github.com/pandas-dev/pandas/blob/7cc4d5375e57277f316c5a8bb17a58ab8003c5f1/pandas/tests/groupby/test_custom_metadata.py#L7): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (117 > 88 characters)
> [Line 39:89](https://github.com/pandas-dev/pandas/blob/7cc4d5375e57277f316c5a8bb17a58ab8003c5f1/pandas/tests/groupby/test_custom_metadata.py#L39): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 88 characters)
> [Line 41:89](https://github.com/pandas-dev/pandas/blob/7cc4d5375e57277f316c5a8bb17a58ab8003c5f1/pandas/tests/groupby/test_custom_metadata.py#L41): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (94 > 88 characters)
> [Line 43:22](https://github.com/pandas-dev/pandas/blob/7cc4d5375e57277f316c5a8bb17a58ab8003c5f1/pandas/tests/groupby/test_custom_metadata.py#L43): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 43:23](https://github.com/pandas-dev/pandas/blob/7cc4d5375e57277f316c5a8bb17a58ab8003c5f1/pandas/tests/groupby/test_custom_metadata.py#L43): [E262](https://duckduckgo.com/?q=pep8%20E262) inline comment should start with '# '
> [Line 44:89](https://github.com/pandas-dev/pandas/blob/7cc4d5375e57277f316c5a8bb17a58ab8003c5f1/pandas/tests/groupby/test_custom_metadata.py#L44): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (99 > 88 characters)
> [Line 47:89](https://github.com/pandas-dev/pandas/blob/7cc4d5375e57277f316c5a8bb17a58ab8003c5f1/pandas/tests/groupby/test_custom_metadata.py#L47): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (108 > 88 characters)
> [Line 53:9](https://github.com/pandas-dev/pandas/blob/7cc4d5375e57277f316c5a8bb17a58ab8003c5f1/pandas/tests/groupby/test_custom_metadata.py#L53): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 68:76](https://github.com/pandas-dev/pandas/blob/7cc4d5375e57277f316c5a8bb17a58ab8003c5f1/pandas/tests/groupby/test_custom_metadata.py#L68): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 68:80](https://github.com/pandas-dev/pandas/blob/7cc4d5375e57277f316c5a8bb17a58ab8003c5f1/pandas/tests/groupby/test_custom_metadata.py#L68): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 69:58](https://github.com/pandas-dev/pandas/blob/7cc4d5375e57277f316c5a8bb17a58ab8003c5f1/pandas/tests/groupby/test_custom_metadata.py#L69): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 77:23](https://github.com/pandas-dev/pandas/blob/7cc4d5375e57277f316c5a8bb17a58ab8003c5f1/pandas/tests/groupby/test_custom_metadata.py#L77): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator

Japanuspus committed:
Fix most PEP8 issues

I don't know how to truncate remaining long lines without loosing information.

Japanuspus committed:
Fix remaining PEP8 issues

Japanuspus committed:
Apply "black" styling and fix typo

This should hopefully resolve remaining linter issues.

Japanuspus committed:
Fix import sorting issues reported by `isort`

Japanuspus commented:
@jbrockmendel This PR is a two-line edit to fix a metadata regression in your recent work on groupby-performance. Would you be available to review? 

I am available to implement this is differently if needed.

jbrockmendel commented on pandas/tests/groupby/test_custom_metadata.py:
```diff
@@ -0,0 +1,89 @@
+
+@pytest.fixture
+def dft():
+    df = pd.DataFrame([[11, 12, 0], [21, 22, 0], [31, 32, 1]], columns={"a", "b", "g"})
```
having columns be a set is really unusual, can you make this a list

Japanuspus replied:
:blush: (converted from a dictionary init rather late in the process...). Is fixed in the updated concise testcase.

jbrockmendel commented on pandas/tests/groupby/test_custom_metadata.py:
```diff
@@ -0,0 +1,89 @@
+    return "+".join(data)
+
+
+class PandasTable(pd.DataFrame):
```
can this be tested without a subclass?  just set _metadata on a regular dataframe?

Japanuspus replied:
I don't think so: The `__finalize__` implementation in NDFrame copies all fields from `self._metadata`, but since self is a new object this would do nothing unless `_metadata` has been set for the class. We could make a test using a monkeypatched `DataFrame` class, but that would not catch the error of using `DataFrame` constructor rather than calling `_constructor`.

I could have a go at a more minimal test where only subclass features are `_metadata` and `_constructor` ?

Japanuspus replied:
Have updated the testcase to use a minimal DataFrame extension. This still catches the original regression.

jbrockmendel commented:
cc @TomAugspurger since this is about `__finalize__` being called.  I'm ambivalent, since this is Technically Correct, but also a place where we're really trying to avoid overhead since these methods get called in a loop.

TomAugspurger commented:
At a glance, this feels like too low of a level to call `__finalize__`. Why can't we call it on the outer level, in whatever function is defining `DataFrameGroupBy.sum()`?

Japanuspus commented:
Thank you both for looking into this! I will start with a more concise testcase.

Should I also have a go at moving the `__finalize__` call to an outer layer, or are either of you looking into that in detail?

Japanuspus committed:
Make testcase more concise

Japanuspus commented:
Regarding performance implications of `__finalize__` in `_chop`

I tried running the benchmark from [GH-34214 ](https://github.com/pandas-dev/pandas/pull/34214), but this seems to hit `.fast_apply` and not `._chop`.

Can either of you come up with a benchmark that hits `._chop` in a hot loop where calling `__finalize__` on the groups would really be wasted?

(`fast_apply` benchmark for reference)
```
import numpy as np
from pandas import DataFrame

N = 10 ** 4
labels = np.random.randint(0, 2000, size=N)
labels2 = np.random.randint(0, 3, size=N)
df = DataFrame(
    {
        "key": labels,
        "key2": labels2,
        "value1": np.random.randn(N),
        "value2": ["foo", "bar", "baz", "qux"] * (N // 4),
    }
)

%timeit df.groupby(["key", "key2"]).apply(lambda x: 1)
%timeit df.groupby("key").apply(lambda x: 1)
```

jbrockmendel commented:
> Can either of you come up with a benchmark that hits ._chop in a hot loop where calling __finalize__ on the groups would really be wasted?

Try setting `df.index = pd.CategoricalIndex(df.index)`

Japanuspus commented:
Just running `df.index = pd.CategoricalIndex(df.index)` for the `df` produced in the `fast_apply` benchmark does not hit `_.chop`. But I have a feeling that was maybe not what you meant?

TomAugspurger commented:
Perhaps run the groupby with a CategoricalIndex?

Japanuspus commented:
Thanks for helping out! Managed to get a benchmark with some degradation:

With `df` as above
```
df.index = pd.CategoricalIndex(df.key)
%timeit df.groupby(level='key').apply(lambda x: 1)
```

Result was a 4% performance hit:
- This patch: 85.3 ms ± 570 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
- 1.1.x from source: 81.9 ms ± 1.48 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

I will not have time to learn enough of the inner workings of pandas to improve this on my own, but am willing to help out if you have some concrete pointers.  
Alternatively you could accept as-is, and then have a look at performance once tests are in place? Or just drop this (I can copy-paste the testcase to the issue thread) -- for my own use case I have implemented a fallback to handle initialization without `__finalize__`, so all is good for me.

jreback commented on pandas/tests/groupby/test_custom_metadata.py:
```diff
@@ -0,0 +1,25 @@
+    custom_df = CustomDataFrame(
+        [[11, 12, 0], [21, 22, 0], [31, 32, 1]], columns=["a", "b", "g"]
+    )
+    custom_df._custom_metadata = "Custom metadata"
```
can you assert that the OP is working as well: https://github.com/pandas-dev/pandas/issues/29442

Japanuspus replied:
Well spotted -- must admit I did not run the code from the original issue since it was missing inputs. After creating some input and adding the test, I realized that the issue I originally fixed was entirely unrelated...

In any case, I have also proposed a solution for the `.sum()` issue as well. Will resolve this comment, please open a new discussion directly on the proposed solution https://github.com/pandas-dev/pandas/pull/35688/commits/8d3d896bcb20f7bcb0d3ac2b71098dfb05fff11e.

If these two issues should be split into two pull requests, please let me know.

Japanuspus committed:
Include test from original issue

Added testcase `test_groupby_sum_with_custom_metadata` for functionality exercised in the #GH-29442.
Testcase fails on current code.

Japanuspus committed:
Fix metadata handling for `.groupby(...).sum()`

In order to propagate metadata fields, the `__finalize__` method must be
called for the resulting DataFrame with a reference to input. By
implementing this in `_GroupBy._agg_general`, this is performed as late
as possible for the `.sum()` (and similar) code-paths.

Fixes #GH-29442

Japanuspus committed:
Apply black format

TomAugspurger commented:
I guess one way to get a sense for the performance impact is to count how many times `__finalize__` is called. I think ideally it's called exactly once, rather than once per group.

Japanuspus commented:
Good point @TomAugspurger.

With respect to my first commit, adding `__finalize__` in `DataSplitter._chop`, it seems this could be improved for the code paths tested here by moving `__finalize__` out to `_GroupBy.__iter__` and `_GroupBy.apply`. Should I make an attempt to do so, or would you rather take another approach? 

With respect to the `sum`-test, I could not see how to move the call  further out than `_agg_general` without having to duplicate it across all the aggregation methods? https://github.com/pandas-dev/pandas/pull/35688/commits/8d3d896bcb20f7bcb0d3ac2b71098dfb05fff11e

Also seems I have a bunch of failing tests, so will take a look at that. Please let me know if you would rather take this in another direction.

simonjayhawkins reviewed (changes requested):
does this need a release note?

Japanuspus commented:
@simonjayhawkins Yes, I should probably add a release note -- but there was a bit of scope creep (my original PR addressed my own problem, not the issue listed), and that work broke something. Also, I have come around to agreeing that we need a more performant version of my initial fix.

I will be short on time for a week or so, but then I will have a go at this.

Japanuspus commented:
Just a note (mostly to self) on the failing tests since it took me an embarrassing amount of time to identify them.

The relevant tests are in `tests/generic/test_finalize.py` and are marked [`xfail`](https://docs.pytest.org/en/reorganize-docs/new-docs/user/xfail.html), which means that when running tests in strict mode, it will be reported as an error if they pass.
Since they include a test for aggregation by sum, they now pass which caused the test to be red in strict mode.

I will remove the `xfail` flag from the relevant cases, and then proceed to look at the performance as discussed above.

Japanuspus committed:
Remove xfail decorator for `sum` aggregation test

jbrockmendel commented:
> The relevant tests are in tests/generic/test_finalize.py and are marked xfail

Tracking down xfailed tests that can be fixed is really helpful, thanks.

Japanuspus committed:
Merge branch 'master' into BUG_GH29442_frame_groupby_metadata

Japanuspus committed:
Move `__finalize__` outsize context handler

Japanuspus committed:
Revert __finalize__ call in DataSplitter

The call was an attempt to ensure that `__finalize__` was called for DataFrames returned when iterating over grouping result, but it caused unreasonable performance degradation.
The relevant test has been marked with `xfail` in order to preserve it. When a solution to the issue is implemented, this test should be reactivated.

Japanuspus committed:
Add performance test counting __finalize__ calls

Japanuspus committed:
Add whatsnew entry

Japanuspus committed:
Move __finalize__ inside context manager

In attempt to address test failure in pandas\tests\plotting\test_groupby.py that I cannot reproduce locally.

Japanuspus committed:
Fix typo in whatsnew entry

Japanuspus commented:
This got sort of messy, and I think a summary is in place:

My initial PR did not address the linked issue and would cause performance regressions. When this was keenly pointed, I added a test and fix for the original issue.

After an unfortunate pause I continued to identify and disable x-failing tests caused by the fix, and also added a test for performance, counting the number of `__finalize__` calls in a `groupby().apply()` invocation, which shares relevant code-path with `.sum()`.

Finally, I have reverted the calls to `__finalize__` from my original PR since this is really an unrelated issue. I have left the test (with an `xfail`-tag), but will file a separate issue for this.

jreback commented on doc/source/whatsnew/v1.1.3.rst:
```diff
@@ -58,6 +58,7 @@ Bug fixes
 - Bug in :meth:`Series.astype` showing too much precision when casting from ``np.float32`` to string dtype (:issue:`36451`)
 - Bug in :meth:`Series.isin` and :meth:`DataFrame.isin` when using ``NaN`` and a row length above 1,000,000 (:issue:`22205`)
 - Bug in :func:`cut` raising a ``ValueError`` when passed a :class:`Series` of labels with ``ordered=False`` (:issue:`36603`)
+- Bug in :class:`BaseGroupedBy` causing ``groupby(...).sum()`` and similar to not preserve metadata (:issue:`29442`)
```
move to 1.1.4

jreback commented on pandas/tests/groupby/test_custom_metadata.py:
```diff
@@ -0,0 +1,81 @@
+
+
+@pytest.mark.xfail(
+    reason="Missing high-performance implementation of metadata support for groupby. "
```
what is the issue here?

Japanuspus replied:
TLDR: I plan to submit a separate PR to lift this x-fail which is unrelated to the linked issue.

The initial commit in this PR (7cc4d5375e57277f316c5a8bb17a58ab8003c5f1) included a call to `__finalize__` in the concrete `DataSplitter`-classes, but as pointed out by @TomAugspurger  and @jbrockmendel , this led to performance-degradation for aggregation code-paths due to unneeded calls to `__finalize__` for the intermediate dataframe/series objects. This is what is tested for in the performance test added in a2bdc3d3dfe80018bb6cabce3583fdf9df6cdca9.

Since calling `__finalize__` in `DataSplitter` turned out to be irrelevant for the linked issue (which was fixed by adding a `__finalize__` call in `BaseGroupBy._agg_general`, see 8d3d896bcb20f7bcb0d3ac2b71098dfb05fff11e) it has been removed from this PR.

One solution to lift the x-fail would be to call `__finalize__` for the groups returned by the iterator returned by `BaseGroup.get_iterator` (see 0af015ba77d8ff05f0d38a3d4b0518bbc737d393 for an implementation on my fork). I am getting some spurious test failures on this branch.

The reason I have not included this solution in the present PR is that I would rather submit it in a separate PR: Calling `__finalize__` independently from object initialization could potentially lead to duplicate calls to `__finalize__` for the same object, and I do not have sufficient overview of the code so see if this is the case in the proposed implementation. My thought was that a separate PR would make it easier to get an overview of such issues -- but I would of course be happy to merge into this PR if you prefer.

jreback reviewed (changes requested):
this looks fine; i doubt performance is actually an issue here, can you show a benchmark otherwise?

Japanuspus committed:
Revert 1.1.3 whatsnew entry

Japanuspus committed:
Merge branch 'master' into BUG_GH29442_frame_groupby_metadata

Japanuspus committed:
Add 1.1.4 whatsnew entry

Japanuspus commented:
> 
> this looks fine; i doubt performance is actually an issue here, can you show a benchmark otherwise?

For the PR as it stands, I don't think performance can be improved.

For the, unrelated and now removed, change in `DataSplitter`  (7cc4d5375e57277f316c5a8bb17a58ab8003c5f1) there is a benchmark in the comments above where we saw a 4% penalty. The alternative solution in 0af015ba77d8ff05f0d38a3d4b0518bbc737d393 only performs the `__finalize__` after aggregation and has no performance issues, but is of course less solid in the sense that there could be code-paths with multiple calls to `__finalize__`. This is the part I propose to move to another PR.

The applied benchmark for reference: 
```
import numpy as np
import pandas as pd

N = 10 ** 4
labels = np.random.randint(0, 2000, size=N)
labels2 = np.random.randint(0, 3, size=N)
df = pd.DataFrame(
    {
        "key": labels,
        "key2": labels2,
        "value1": np.random.randn(N),
        "value2": ["foo", "bar", "baz", "qux"] * (N // 4),
    }
)
df.index = pd.CategoricalIndex(df.key)
%timeit df.groupby(level='key').apply(lambda x: 1)

```

Japanuspus commented:
> 
> 
> does this need a release note?

Added a release note for 1.1.4

jreback commented:
@Japanuspus pls limit changes to one thing in a PR, esp as this is a backport.

Japanuspus committed:
Revert all changes not strictly related to issue

Japanuspus committed:
Use descriptive names for unit tests

Japanuspus committed:
Remove extraneous imports

Japanuspus committed:
Merge master into BUG_GH29442_frame_groupby_metadata

Merging from commit 'ec8c1c4ecf1f0d375d8d1287ee5cbd456852faea' which is most recent commit with green tests

jreback commented on doc/source/whatsnew/v1.1.4.rst:
```diff
@@ -22,7 +22,7 @@ Fixed regressions
 Bug fixes
 ~~~~~~~~~
--
+- Bug in :class:`BaseGroupedBy` causing ``groupby(...).sum()`` and similar to not preserve metadata (:issue:`29442`)
```
you can remove the reference to BaseGroupedBy (this is not a real class) and these are user facing notes.

jreback commented on pandas/tests/groupby/test_custom_metadata.py:
```diff
@@ -0,0 +1,66 @@
+        {
+            "key": labels,
+            "key2": labels2,
+            "value1": np.random.randn(N),
```
this file can be removed (as its not germane to the bug fix right)? if its a followon that's fine, but do in a separate PR targeting master.

Japanuspus replied:
The key test was a regression test for the linked issue, but there should be full coverage even without this test.

Japanuspus committed:
Delete tests for custom metadata

Japanuspus committed:
Remove reference to internal class from whatsnew

Japanuspus commented:
I am slightly confused about @jreback mentioning this being a backport PR -- my intention was to target master. Or does this just mean that the patch will both apply to master and release branches?

jreback commented:
> I am slightly confused about @jreback mentioning this being a backport PR -- my intention was to target master. Or does this just mean that the patch will both apply to master and release branches?

this si a regresion, so we target master, then its backported. so pls limit the change to the regression. if you want to do another change that targets master only that's ok, but separate PR

Japanuspus commented:
> > I am slightly confused about @jreback mentioning this being a backport PR -- my intention was to target master. Or does this just mean that the patch will both apply to master and release branches?
> 
> this si a regresion, so we target master, then its backported. so pls limit the change to the regression. if you want to do another change that targets master only that's ok, but separate PR

Thank you for the explanation.
I believe the current PR is cut as tight as possible -- will go ahead with a separate PR for the `iter(...groupby())` issue.

jreback commented:
lgtm. @TomAugspurger if any comments.

jreback commented:
thanks @Japanuspus
